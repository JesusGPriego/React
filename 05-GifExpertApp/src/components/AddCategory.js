import React from "react";
import { useState } from "react";
//proptypes
import PropTypes from "prop-types";

//getting the funciton (allowing component comunication)
/**
 * Add category allow the user to add more categories to the category list.
 * it is a simple input text, that add another category to the listo when 'enter' is pushed.
 * @param {setCategories} is the function that is triggered whenever enter is pushed. Another category is added so, another grid item will be created.
 */
export const AddCategory = ({ setCategories }) => {
  const [inputValue, setinputValue] = useState(""); //not leave empty, or undefined is set.

  const handleInputChange = (e) => {
    // console.log(e.target.value);
    //input value is all the text written, so by setting this value to input value, we actually got the input value.
    setinputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    // console.log("Submitted");
    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setinputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Insert new category"
      />
    </form>
  );
};

AddCategory.propTypes = {
  //setting the function as required:
  setCategories: PropTypes.func.isRequired,
};
