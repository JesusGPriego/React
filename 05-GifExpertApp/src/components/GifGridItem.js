import React from "react";
import PropTypes from "prop-types";
/**
 * Creates a grid item for the GifGrid. 
 * This item display an image and its title.
 * 
 * @param {id} is the image id.
 * @param {title} it is the image title.
 * @param {url} it is the image's url.
 * @returns a card that contains an image, and the image's title.
 */
export const GifGridItem = ({ title, url }) => {
  return (
    <figure className="card animate__animated animate__fadeIn">
      <img src={url} alt={title} />
      <figcaption>{title}</figcaption>
    </figure>
  );
};


GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}