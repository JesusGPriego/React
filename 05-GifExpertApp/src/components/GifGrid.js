import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
/**
 * Creates a grid of images that match the topic that the user is looking for.
 * Topic here is the category.
 * 
 * @param {category} is what the user is looking for.  
 * @returns a new fragment, which is a grid with 10 images.
 */
export const GifGrid = ({ category }) => {
  //asking custom hook for info:
  const {data:images, loading} = useFetchGifs(category);

  return (
    <>
    {loading && <p>Cargando...</p>}
      <h3>{category}</h3>
      <div className="card-grid animate__animated animate__fadeIn">
        {images.map((image) => (
          <GifGridItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
