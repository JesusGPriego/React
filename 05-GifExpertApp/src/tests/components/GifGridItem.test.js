import { shallow } from "enzyme";
import {GifGridItem} from '../../components/GifGridItem'


describe('Testing GifGridItem', () => {
    
    const title = 'title';
    const url = 'https://localhost.algo.jpg';
    let wrapper  = shallow(<GifGridItem title={title} url={url}/>)
    

    test('should return GfiGridItem correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should return a <p></p> displayin the title', () => {
        const figcaption = wrapper.find('figcaption');
        expect(figcaption.text().trim()).toBe(title);
    })

    test('should get img=url and alt = title', () => {
        
        const img = wrapper.find('img');
        expect(img.prop('src')).toBe(url);
        expect(img.prop('alt')).toBe(title);

    })
    
    test('should include animate__fadeIn classnName', () => {
        
        const div = wrapper.find('figure');
        expect(div.prop('className')).toContain('animate__fadeIn');

    })
    

})
