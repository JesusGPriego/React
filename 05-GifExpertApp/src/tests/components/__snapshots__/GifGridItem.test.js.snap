// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGridItem should return GfiGridItem correctly 1`] = `
<figure
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="title"
    src="https://localhost.algo.jpg"
  />
  <figcaption>
    title
  </figcaption>
</figure>
`;
